File: cpp/src/processor.cpp
#include "processor.h"

#include <istream>
#include <ostream>
#include <sstream>
#include <utility>

#include "iban.h"
#include "parse_error.h"

namespace settle {
namespace {

constexpr std::string_view kSupportedCurrency = "EUR";
constexpr Cents kRtgsThreshold = 1'000'000;  // 10,000.00

std::vector<std::string> split(const std::string& line, char separator) {
    std::vector<std::string> fields;
    std::istringstream stream(line);
    std::string field;
    while (std::getline(stream, field, separator)) fields.push_back(field);
    if (!line.empty() && line.back() == separator) fields.emplace_back();
    return fields;
}

bool parseFlag(const std::string& text) {
    if (text == "true") return true;
    if (text == "false") return false;
    throw ParseError("urgent must be true or false: '" + text + "'");
}

}  // namespace

BatchProcessor::BatchProcessor() {
    rails_.push_back(std::make_unique<BookRail>());
    rails_.push_back(std::make_unique<AchRail>());
    rails_.push_back(std::make_unique<RtgsRail>());
    for (const auto& rail : rails_) totals_[rail->name()] = RailTotals{};
}

void BatchProcessor::run(std::istream& input, std::ostream& output) {
    output << "== Payments ==\n";

    std::string line;
    bool header = true;
    while (std::getline(input, line)) {
        if (!line.empty() && line.back() == '\r') line.pop_back();  // a file written on Windows
        if (header) {
            header = false;
            continue;
        }
        if (line.empty()) continue;
        process(line, output);
    }

    output << "== Rails ==\n";
    for (const auto& rail : rails_) {
        const RailTotals& totals = totals_[rail->name()];
        output << rail->name() << " count=" << totals.count
               << " volume=" << formatAmount(totals.volume)
               << " fees=" << formatAmount(totals.fees) << '\n';
    }

    output << "== Positions ==\n";
    Cents sum = 0;
    for (const auto& [account, cents] : positions_) {
        output << account << ' ' << formatAmount(cents) << '\n';
        sum += cents;
    }

    output << "== Check ==\n"
           << "accepted=" << accepted_ << " rejected=" << rejected_
           << " sum=" << formatAmount(sum) << (sum == 0 ? " OK" : " MISMATCH") << '\n';
}

void BatchProcessor::process(const std::string& line, std::ostream& output) {
    const std::vector<std::string> fields = split(line, ',');

    Payment payment;
    try {
        payment = readPayment(fields);
    } catch (const ParseError&) {
        reject(fields.empty() ? std::string_view{} : std::string_view{fields[0]}, "MALFORMED", output);
        return;
    }

    if (const std::string_view reason = check(payment); !reason.empty()) {
        reject(payment.id, reason, output);
        return;
    }

    const Rail& rail = route(payment);
    const Cents fee = rail.fee(payment.amount);
    post(payment, fee);

    RailTotals& totals = totals_[rail.name()];
    ++totals.count;
    totals.volume += payment.amount;
    totals.fees += fee;

    ++accepted_;
    output << payment.id << " ACCEPTED " << rail.name() << ' ' << formatAmount(payment.amount)
           << " fee=" << formatAmount(fee) << '\n';
}

Payment BatchProcessor::readPayment(const std::vector<std::string>& fields) const {
    if (fields.size() != 6) {
        throw ParseError("expected 6 fields, got " + std::to_string(fields.size()));
    }

    Payment payment;
    payment.id = fields[0];
    payment.debtor = fields[1];
    payment.creditor = fields[2];
    payment.amount = parseAmount(fields[3]);
    payment.currency = fields[4];
    payment.urgent = parseFlag(fields[5]);
    return payment;
}

// The first rule that fails, or an empty view when the payment is good.
std::string_view BatchProcessor::check(const Payment& payment) {
    if (!seenIds_.insert(payment.id).second) return "DUPLICATE_ID";
    if (!isValidIban(payment.debtor) || !isValidIban(payment.creditor)) return "INVALID_IBAN";
    if (payment.debtor == payment.creditor) return "SAME_ACCOUNT";
    if (payment.amount <= 0) return "INVALID_AMOUNT";
    if (payment.currency != kSupportedCurrency) return "UNSUPPORTED_CURRENCY";
    return {};
}

const Rail& BatchProcessor::route(const Payment& payment) const {
    // Same bank: the money never leaves our books, whatever the sender asked for.
    if (bankCode(payment.debtor) == bankCode(payment.creditor)) return *rails_[0];
    if (payment.urgent || payment.amount >= kRtgsThreshold) return *rails_[2];
    return *rails_[1];
}

void BatchProcessor::reject(std::string_view id, std::string_view reason, std::ostream& output) {
    output << id << " REJECTED " << reason << '\n';
    ++rejected_;
}

// Double entry: the sender pays amount plus fee, the beneficiary receives the amount,
// and FEES takes the difference, so the batch nets to exactly zero.
void BatchProcessor::post(const Payment& payment, Cents fee) {
    positions_[payment.debtor] -= payment.amount + fee;
    positions_[payment.creditor] += payment.amount;
    positions_["FEES"] += fee;
}

}  // namespace settle

File: go/processor.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"sort"
	"strings"
)

const (
	supportedCurrency = "EUR"
	rtgsThreshold     = 1_000_000 // 10,000.00
	feeAccount        = "FEES"
)

// errParse marks a line that cannot be read as a payment at all, as opposed to one
// that fails a business rule. Go has no exceptions: errors are values, and callers
// match on them with errors.Is.
var errParse = errors.New("parse error")

type payment struct {
	id       string
	debtor   string
	creditor string
	amount   int64
	currency string
	urgent   bool
}

type railTotals struct {
	count  int
	volume int64
	fees   int64
}

// processor reads a clearing file, applies the scheme rules, routes each accepted
// payment and prints the report. Rejection reasons are checked in a fixed order and
// the first failure wins, so a line always has exactly one reason.
type processor struct {
	rails     []Rail
	totals    map[string]*railTotals
	seenIDs   map[string]bool
	positions map[string]int64
	accepted  int
	rejected  int
}

func newProcessor() *processor {
	p := &processor{
		rails:     []Rail{bookRail{}, achRail{}, rtgsRail{}},
		totals:    make(map[string]*railTotals),
		seenIDs:   make(map[string]bool),
		positions: make(map[string]int64),
	}
	for _, rail := range p.rails {
		p.totals[rail.Name()] = &railTotals{}
	}
	return p
}

func (p *processor) run(input io.Reader, output io.Writer) error {
	if _, err := fmt.Fprintln(output, "== Payments =="); err != nil {
		return err
	}

	scanner := bufio.NewScanner(input)
	header := true
	for scanner.Scan() {
		line := strings.TrimSuffix(scanner.Text(), "\r") // a file written on Windows
		if header {
			header = false
			continue
		}
		if line == "" {
			continue
		}
		p.process(line, output)
	}
	if err := scanner.Err(); err != nil {
		return err
	}

	fmt.Fprintln(output, "== Rails ==")
	for _, rail := range p.rails {
		t := p.totals[rail.Name()]
		fmt.Fprintf(output, "%s count=%d volume=%s fees=%s\n",
			rail.Name(), t.count, formatAmount(t.volume), formatAmount(t.fees))
	}

	fmt.Fprintln(output, "== Positions ==")
	accounts := make([]string, 0, len(p.positions))
	for account := range p.positions {
		accounts = append(accounts, account)
	}
	sort.Strings(accounts) // map iteration order is random in Go, so sort explicitly

	var sum int64
	for _, account := range accounts {
		fmt.Fprintf(output, "%s %s\n", account, formatAmount(p.positions[account]))
		sum += p.positions[account]
	}

	status := "OK"
	if sum != 0 {
		status = "MISMATCH"
	}
	fmt.Fprintln(output, "== Check ==")
	_, err := fmt.Fprintf(output, "accepted=%d rejected=%d sum=%s %s\n",
		p.accepted, p.rejected, formatAmount(sum), status)
	return err
}

func (p *processor) process(line string, output io.Writer) {
	fields := strings.Split(line, ",")

	pmt, err := readPayment(fields)
	if err != nil {
		id := ""
		if len(fields) > 0 {
			id = fields[0]
		}
		p.reject(id, "MALFORMED", output)
		return
	}

	if reason := p.check(pmt); reason != "" {
		p.reject(pmt.id, reason, output)
		return
	}

	rail := p.route(pmt)
	fee := rail.Fee(pmt.amount)
	p.post(pmt, fee)

	t := p.totals[rail.Name()]
	t.count++
	t.volume += pmt.amount
	t.fees += fee

	p.accepted++
	fmt.Fprintf(output, "%s ACCEPTED %s %s fee=%s\n",
		pmt.id, rail.Name(), formatAmount(pmt.amount), formatAmount(fee))
}

func readPayment(fields []string) (payment, error) {
	if len(fields) != 6 {
		return payment{}, fmt.Errorf("%w: expected 6 fields, got %d", errParse, len(fields))
	}

	amount, err := parseAmount(fields[3])
	if err != nil {
		return payment{}, err
	}

	urgent, err := parseFlag(fields[5])
	if err != nil {
		return payment{}, err
	}

	return payment{
		id:       fields[0],
		debtor:   fields[1],
		creditor: fields[2],
		amount:   amount,
		currency: fields[4],
		urgent:   urgent,
	}, nil
}

func parseFlag(text string) (bool, error) {
	switch text {
	case "true":
		return true, nil
	case "false":
		return false, nil
	default:
		return false, fmt.Errorf("%w: urgent must be true or false: %q", errParse, text)
	}
}

// check returns the first rule that fails, or "" when the payment is good.
func (p *processor) check(pmt payment) string {
	if p.seenIDs[pmt.id] {
		return "DUPLICATE_ID"
	}
	p.seenIDs[pmt.id] = true

	if !isValidIban(pmt.debtor) || !isValidIban(pmt.creditor) {
		return "INVALID_IBAN"
	}
	if pmt.debtor == pmt.creditor {
		return "SAME_ACCOUNT"
	}
	if pmt.amount <= 0 {
		return "INVALID_AMOUNT"
	}
	if pmt.currency != supportedCurrency {
		return "UNSUPPORTED_CURRENCY"
	}
	return ""
}

func (p *processor) route(pmt payment) Rail {
	// Same bank: the money never leaves our books, whatever the sender asked for.
	if bankCode(pmt.debtor) == bankCode(pmt.creditor) {
		return p.rails[0]
	}
	if pmt.urgent || pmt.amount >= rtgsThreshold {
		return p.rails[2]
	}
	return p.rails[1]
}

func (p *processor) reject(id, reason string, output io.Writer) {
	fmt.Fprintf(output, "%s REJECTED %s\n", id, reason)
	p.rejected++
}

// post writes the double entry: the sender pays amount plus fee, the beneficiary
// receives the amount, and FEES takes the difference, so the batch nets to zero.
func (p *processor) post(pmt payment, fee int64) {
	p.positions[pmt.debtor] -= pmt.amount + fee
	p.positions[pmt.creditor] += pmt.amount
	p.positions[feeAccount] += fee
}
